app/controllers/application_controller.rb

```diff
@@ -5,6 +5,143 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :sessions_secret, "secret"
   end
 
-end
\ No newline at end of file
+# Users Actions
+  get '/' do
+    erb :index
+  end
+
+  get '/login' do
+    if Helpers.is_logged_in?(session)
+      redirect '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  get '/signup' do
+    if Helpers.is_logged_in?(session)
+
+      redirect '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if !(params.any? {|key, userdata| userdata.empty?})
+      @user = User.create(params)
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/logout' do
+    if Helpers.is_logged_in?(session)
+      session.clear
+      redirect '/login'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    session[:id] = user.id
+    redirect '/tweets'
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+#tweets actions
+  get '/tweets' do
+    if Helpers.is_logged_in?(session)
+      @user = User.find_by_id(session[:id])
+      erb :'tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if Helpers.is_logged_in?(session)
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/new' do
+    @tweet = Tweet.create(content: params[:content], user_id: session[:id])
+    if @tweet.content ==""
+        redirect "/tweets/new"
+    else
+      redirect "/tweets/#{@tweet.id}"
+
+    end
+  end
+
+  post '/tweets' do
+    @user = User.find_by_id(params[:id])
+    erb :'tweets/tweets'
+  end
+
+  get '/tweets/:id' do
+    if Helpers.is_logged_in?(session)
+      @tweet = Tweet.find_by_id(params[:id])
+        erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+
+  end
+
+  patch '/tweets/:id/edit' do
+    @tweet = Tweet.find_by_id(params[:id])
+    @tweet.content = params[:content]
+    @tweet.save
+    if !(@tweet.content)
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if Helpers.is_logged_in?(session)
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/edit_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == session[:id]
+        @tweet.delete
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.